src/transform/ir_proto_to_datalog/arcs_manifest_tree/type.h

```diff
@@ -0,0 +1,18 @@
+#ifndef SRC_TRANSFORM_IR_PROTO_TO_DATALOG_ARCS_MANIFEST_TREE_TYPE_H_
+#define SRC_TRANSFORM_IR_PROTO_TO_DATALOG_ARCS_MANIFEST_TREE_TYPE_H_
+
+#include "absl/container/flat_hash_set.h"
+#include "src/ir/access_path_selectors_set.h"
+#include "third_party/arcs/proto/manifest.pb.h"
+
+namespace raksha::transform::arcs_manifest_tree {
+
+class Type {
+ public:
+  virtual ~Type() {}
```

Comment:
We can't use `absl::variant` here because `Type`s recursively nest within `Type`s. `absl::variant` acts like a `union` with an `enum` describing which variant we are currently using. Because `EntityType`s contain an embedded map of field names to field `Type`s, using `absl::variant` turns into asking the compiler to embed a `Type` within a `Type`. This is impossible, as C++ class objects are of fixed width with fields at fixed offsets from the top of the object. Therefore, we have to introduce a pointer to point at the nested type.

Because we have to introduce a pointer, we have to deal with lifetime concerns anyway. The benefits of `absl::variant` have disappeared, and we can more cheaply and correctly use regular ol' C++ inheritance.